src/middlewares/languageMiddleware.ts

```diff
@@ -0,0 +1,179 @@
+ * - given the following valid pages, an English user will be redirected from (1) to (2), and not to (3):
+ *      - (1) `/gift`
+ *      - (2) `/en/poison`
+ *      - (3) `/en/gift`
+ * - given the following valid pages, an English user navigating to (1) will not be redirected to (2), but receive a 404 error:
+ *      - (1) `/gift`
+ *      - (2) `/en/gift`
+ *
+ * @param request
+ */
+export async function languageMiddleware(
+  request: NextRequest,
+): Promise<NextResponse | undefined> {
+  const availableLanguages = await client.fetch<LanguageObject[] | null>(
+    LANGUAGES_QUERY,
+  );
+  if (availableLanguages === null) {
+    console.error("No languages available, language middleware aborted.");
+    return;
+  }
+  const pathname = request.nextUrl.pathname;
+  const language = availableLanguages.find(
+    ({ id }) => pathname.startsWith(`/${id}/`) || pathname === `/${id}`,
+  );
+  if (language === undefined) {
+    return redirectMissingLanguage(pathname, availableLanguages, request.url);
+  }
+  return redirectWithLanguage(pathname, language, request.url);
+}
+
+/**
+ * Check that the slug actually exists for the given language.
+ * - If it exists, no changes are made.
+ * - Otherwise, we attempt to translate to the specified language
+ *   - If translated, user is redirected to `/[language]/[translatedSlug]`
+ *   - Otherwise, user is redirected to the slug with default language (`/[slug]`)
+ *
+ * @param pathname
+ * @param language
+ * @param baseUrl
+ */
+async function redirectWithLanguage(
+  pathname: string,
+  language: LanguageObject,
+  baseUrl: string,
+) {
+  const slug = pathname.split("/").slice(2).join("/");
+  const translatedSlug = await translateSlug(slug, language.id);
+  if (translatedSlug === undefined) {
+    console.warn("Redirect", pathname, "→", `/${slug}`);
+    return NextResponse.redirect(new URL(`/${slug}`, baseUrl));
+  }
+  if (translatedSlug !== slug) {
+    const translatedPathname = `/${language.id}/${translatedSlug}`;
+    console.warn("Redirect", pathname, "→", translatedPathname);
+    return NextResponse.redirect(new URL(translatedPathname, baseUrl));
+  }
+  // all good, no modifications needed
+  return;
+}
+
+/**
+ *
+ * Negotiate with the user to find the most preferred languages from the list of available languages
+ *   - If the negotiated language is the same as default, no changes are made.
+ *   - Otherwise, we attempt to translate to the negotiated language
+ *     - If translated, user is redirected to `/[negotiatedLanguage]/[translatedSlug]`
+ *     - Otherwise, user is redirected to the slug with default language (`/[slug]`)
+ *
+ * @param pathname
+ * @param availableLanguages
+ * @param baseUrl
+ */
+async function redirectMissingLanguage(
+  pathname: string,
+  availableLanguages: LanguageObject[],
+  baseUrl: string,
+) {
+  // Pathname does NOT include a language code, we negotiate with the user to find
+  // the most preferred languages from the list of available languages
```

Comment:
How will this interaction work?

Reply:
See `negotiateClientLanguage`: The `Accept-Language` header is used to read the browser language list, and the top supported language is picked (or returns `undefined`).